File: scripts/validate_underscore_markdown.py
import json
import re
from pathlib import Path
import sys

ROOT = Path(__file__).resolve().parent.parent
LOG_PATH = ROOT / ".cursor" / "debug-3b0766.log"
SESSION_ID = "3b0766"
RUN_ID = sys.argv[1] if len(sys.argv) > 1 else "pre-fix"


def log_event(hypothesis_id, location, message, data):
    payload = {
        "sessionId": SESSION_ID,
        "runId": RUN_ID,
        "hypothesisId": hypothesis_id,
        "location": location,
        "message": message,
        "data": data,
        "timestamp": int(__import__("time").time() * 1000),
    }
    # region agent log
    LOG_PATH.parent.mkdir(parents=True, exist_ok=True)
    with LOG_PATH.open("a", encoding="utf-8") as handle:
        handle.write(json.dumps(payload, ensure_ascii=True) + "\n")
    # endregion


def collect_underscore_markdown_files(base):
    files = []
    for path in base.rglob("*.md"):
        rel_parts = path.relative_to(base).parts
        if any(part.startswith("_") for part in rel_parts):
            files.append(path.relative_to(base).as_posix())
    return sorted(files)


def has_frontmatter(content):
    return bool(re.match(r"^---\s*\n[\s\S]*?\n---\s*(\n|$)", content))


def has_html(content):
    return bool(re.search(r"</?[A-Za-z][^>\n]*>", content))


def has_non_ascii(content):
    return any(ord(char) > 127 for char in content)


rule_path = ROOT / ".cursor" / "rules" / "underscore-markdown-plain-text.mdc"
rule_content = rule_path.read_text(encoding="utf-8") if rule_path.exists() else ""
has_globs_line = bool(re.search(r"^globs:\s+.+", rule_content, re.MULTILINE))

# region agent log
log_event(
    "H1",
    "scripts/validate_underscore_markdown.py:52",
    "Rule file globs check",
    {"rulePath": str(rule_path.relative_to(ROOT)), "hasGlobsLine": has_globs_line},
)
# endregion

underscore_markdown_files = collect_underscore_markdown_files(ROOT)

# region agent log
log_event(
    "H4",
    "scripts/validate_underscore_markdown.py:64",
    "Underscore markdown discovery",
    {"count": len(underscore_markdown_files), "files": underscore_markdown_files},
)
# endregion

frontmatter_violations = []
html_violations = []
non_ascii_violations = []

for rel_path in underscore_markdown_files:
    content = (ROOT / rel_path).read_text(encoding="utf-8")
    if has_frontmatter(content):
        frontmatter_violations.append(rel_path)
    if has_html(content):
        html_violations.append(rel_path)
    if has_non_ascii(content):
        non_ascii_violations.append(rel_path)

# region agent log
log_event(
    "H2",
    "scripts/validate_underscore_markdown.py:84",
    "Frontmatter validation",
    {
        "violationCount": len(frontmatter_violations),
        "violations": frontmatter_violations,
    },
)
# endregion

# region agent log
log_event(
    "H3",
    "scripts/validate_underscore_markdown.py:96",
    "Plain markdown validation",
    {
        "htmlViolationCount": len(html_violations),
        "htmlViolations": html_violations,
        "nonAsciiViolationCount": len(non_ascii_violations),
        "nonAsciiViolations": non_ascii_violations,
    },
)
# endregion

langsmith_file = next(
    (file for file in underscore_markdown_files if "langsmith" in file.lower()), None
)
langflow_file = next(
    (file for file in underscore_markdown_files if "langflow" in file.lower()), None
)
langsmith_has_langgraph = (
    bool(re.search(r"langgraph", (ROOT / langsmith_file).read_text(encoding="utf-8"), re.I))
    if langsmith_file
    else False
)
langflow_has_langgraph = (
    bool(re.search(r"langgraph", (ROOT / langflow_file).read_text(encoding="utf-8"), re.I))
    if langflow_file
    else False
)

# region agent log
log_event(
    "H5",
    "scripts/validate_underscore_markdown.py:126",
    "LangGraph context coverage",
    {
        "langsmithFile": langsmith_file,
        "langflowFile": langflow_file,
        "langsmithHasLangGraph": langsmith_has_langgraph,
        "langflowHasLangGraph": langflow_has_langgraph,
    },
)
# endregion

checks_passed = (
    has_globs_line
    and len(underscore_markdown_files) >= 2
    and len(frontmatter_violations) == 0
    and len(html_violations) == 0
    and len(non_ascii_violations) == 0
    and bool(langsmith_file)
    and bool(langflow_file)
    and langsmith_has_langgraph
    and langflow_has_langgraph
)

# region agent log
log_event(
    "H-SUMMARY",
    "scripts/validate_underscore_markdown.py:151",
    "Overall validation summary",
    {"checksPassed": checks_passed},
)
# endregion

File: scripts/validate-underscore-markdown.js
const fs = require("fs");
const path = require("path");

const ROOT = path.resolve(__dirname, "..");
const RUN_ID = process.argv[2] || "pre-fix";
const ENDPOINT = "http://127.0.0.1:7282/ingest/963f32de-fcc8-4211-bb99-d1f1d6605248";
const SESSION_ID = "3b0766";

function collectUnderscoreMarkdownFiles(dirPath, files = []) {
  const entries = fs.readdirSync(dirPath, { withFileTypes: true });
  for (const entry of entries) {
    if (entry.name === ".git" || entry.name === "node_modules") {
      continue;
    }

    const absPath = path.join(dirPath, entry.name);
    if (entry.isDirectory()) {
      collectUnderscoreMarkdownFiles(absPath, files);
      continue;
    }

    if (!entry.isFile() || !entry.name.endsWith(".md")) {
      continue;
    }

    const relPath = path.relative(ROOT, absPath);
    const hasUnderscoreSegment = relPath
      .split(path.sep)
      .some((segment) => segment.startsWith("_"));

    if (hasUnderscoreSegment) {
      files.push(relPath);
    }
  }
  return files;
}

function hasYamlFrontmatter(content) {
  return /^---\s*\n[\s\S]*?\n---\s*(\n|$)/.test(content);
}

function hasHtmlTag(content) {
  return /<\/?[A-Za-z][^>\n]*>/.test(content);
}

function hasNonAscii(content) {
  return /[^\x00-\x7F]/.test(content);
}

function readIfExists(absPath) {
  if (!fs.existsSync(absPath)) {
    return "";
  }
  return fs.readFileSync(absPath, "utf8");
}

function postDebugLog(payload) {
  // #region agent log
  fetch(ENDPOINT, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "X-Debug-Session-Id": "3b0766"
    },
    body: JSON.stringify(payload)
  }).catch(() => {});
  // #endregion
}

const rulePath = path.join(
  ROOT,
  ".cursor",
  "rules",
  "underscore-markdown-plain-text.mdc"
);
const ruleContent = readIfExists(rulePath);
const hasGlobsLine = /^globs:\s+.+/m.test(ruleContent);

// #region agent log
postDebugLog({
  sessionId: SESSION_ID,
  runId: RUN_ID,
  hypothesisId: "H1",
  location: "scripts/validate-underscore-markdown.js:72",
  message: "Rule file globs check",
  data: { rulePath: path.relative(ROOT, rulePath), hasGlobsLine },
  timestamp: Date.now()
});
// #endregion

const underscoreMarkdownFiles = collectUnderscoreMarkdownFiles(ROOT);

// #region agent log
postDebugLog({
  sessionId: SESSION_ID,
  runId: RUN_ID,
  hypothesisId: "H4",
  location: "scripts/validate-underscore-markdown.js:88",
  message: "Underscore markdown discovery",
  data: {
    count: underscoreMarkdownFiles.length,
    files: underscoreMarkdownFiles
  },
  timestamp: Date.now()
});
// #endregion

const frontmatterViolations = [];
const htmlViolations = [];
const nonAsciiViolations = [];

for (const relPath of underscoreMarkdownFiles) {
  const absPath = path.join(ROOT, relPath);
  const content = fs.readFileSync(absPath, "utf8");

  if (hasYamlFrontmatter(content)) {
    frontmatterViolations.push(relPath);
  }
  if (hasHtmlTag(content)) {
    htmlViolations.push(relPath);
  }
  if (hasNonAscii(content)) {
    nonAsciiViolations.push(relPath);
  }
}

// #region agent log
postDebugLog({
  sessionId: SESSION_ID,
  runId: RUN_ID,
  hypothesisId: "H2",
  location: "scripts/validate-underscore-markdown.js:118",
  message: "Frontmatter validation",
  data: {
    violations: frontmatterViolations,
    violationCount: frontmatterViolations.length
  },
  timestamp: Date.now()
});
// #endregion

// #region agent log
postDebugLog({
  sessionId: SESSION_ID,
  runId: RUN_ID,
  hypothesisId: "H3",
  location: "scripts/validate-underscore-markdown.js:130",
  message: "Plain markdown validation",
  data: {
    htmlViolations,
    nonAsciiViolations,
    htmlViolationCount: htmlViolations.length,
    nonAsciiViolationCount: nonAsciiViolations.length
  },
  timestamp: Date.now()
});
// #endregion

const langsmithFile = underscoreMarkdownFiles.find((filePath) =>
  /langsmith/i.test(filePath)
);
const langflowFile = underscoreMarkdownFiles.find((filePath) =>
  /langflow/i.test(filePath)
);
const langsmithHasLangGraph = langsmithFile
  ? /langgraph/i.test(fs.readFileSync(path.join(ROOT, langsmithFile), "utf8"))
  : false;
const langflowHasLangGraph = langflowFile
  ? /langgraph/i.test(fs.readFileSync(path.join(ROOT, langflowFile), "utf8"))
  : false;

// #region agent log
postDebugLog({
  sessionId: SESSION_ID,
  runId: RUN_ID,
  hypothesisId: "H5",
  location: "scripts/validate-underscore-markdown.js:157",
  message: "LangGraph context coverage",
  data: {
    langsmithFile: langsmithFile || null,
    langflowFile: langflowFile || null,
    langsmithHasLangGraph,
    langflowHasLangGraph
  },
  timestamp: Date.now()
});
// #endregion

const checksPassed =
  hasGlobsLine &&
  underscoreMarkdownFiles.length >= 2 &&
  frontmatterViolations.length === 0 &&
  htmlViolations.length === 0 &&
  nonAsciiViolations.length === 0 &&
  Boolean(langsmithFile) &&
  Boolean(langflowFile) &&
  langsmithHasLangGraph &&
  langflowHasLangGraph;

// #region agent log
postDebugLog({
  sessionId: SESSION_ID,
  runId: RUN_ID,
  hypothesisId: "H-SUMMARY",
  location: "scripts/validate-underscore-markdown.js:183",
  message: "Overall validation summary",
  data: { checksPassed },
  timestamp: Date.now()
});
// #endregion

